### PR TITLE
[framework] improve token standard

### PR DESCRIPTION
**Pull request by davidiw**
Huge step forward now that we can seamlessly look up a token data or
token balance with a single query!

* move from ID from GUID to TokenId { creators addr, collection, token }
* add events
* add initial framework for capabilities
* add maximum number of tokens and ability to mint more
* add supply tracking to tokens
* re-organize code to be more data focused
* update SDKs and tutorial code -- likely need to revise tutorials -- but let's land this first...
* updated smoke test
* removed token metadata for now, for simplicity

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/Token.move [hunk 1]
```diff
@@ -1,37 +1,82 @@
...
-    struct Collections has key {
-        collections: Table<ASCII::String, Collection>,
+    /// Represents a unique identity for the token
+    struct TokenId has copy, drop, store {
+        // The creator of this token
+        creator: address,
+        // The collection or set of related tokens within the creator's account
+        collection: ASCII::String,
+        // Unique name within a collection within the creator's account
+        name: ASCII::String,
```
Comment (CapCap) [inline aptos-move/framework/aptos-framework/sources/Token.move:44]: Are you expecting this to be a human readable name for both collection and the NFT name?

Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Token.move:44]: Yes. This part is identical to the previous token.

**Comment by MartianSiddharth** [comment]
> add events
Changes related to events seem to be missing

File: aptos-move/framework/aptos-framework/sources/Token.move [hunk 2]
```diff
@@ -44,362 +89,510 @@ module AptosFramework::Token {
...
-            maximum,
-        };
+        direct_deposit(account_addr, token)
+    }
 
-        Table::add(collections, &name, collection);
+    /// Deposit the token balance into the recipients account and emit an event.
+    public fun direct_deposit(account_addr: address, token: Token) acquires TokenStore {
+        let token_store = borrow_global_mut<TokenStore>(account_addr);
+        Event::emit_event<DepositEvent>(
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/Token.move:246]: @MartianSiddharth

File: aptos-move/framework/aptos-framework/sources/Token.move [hunk 3]
```diff
@@ -44,362 +89,510 @@ module AptosFramework::Token {
...
-            },
-        )
+    public fun withdraw_token(
+        account: &signer,
+        id: &TokenId,
+        amount: u64,
+    ): Token acquires TokenStore {
+        let account_addr = Signer::address_of(account);
+        let token_store = borrow_global_mut<TokenStore>(account_addr);
+        Event::emit_event<WithdrawEvent>(
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/Token.move:335]: @MartianSiddharth

**Comment by davidiw** [comment]
> > add events
> 
> Changes related to events seem to be missing

Commented where they are. Probably the diff was hidden.

**Review by zekun000 — COMMENTED** [review]
just copy some of offline discussion here, I like the ability to group tokens in a single table<TokenId, Token>, the major concern of this is about type safety as we discussed.

the alternative way of doing this is by generics, and treating collection as Type, token as instance of one Type.
```
module NFTStarndard {
struct Token<T> {
  data: T
}
fun transfer<T>
fun mint<T>
...
```

```
module BAYC {
  struct BAYC {
    index: u256
  } 
}
module PUNK {
  struct PUNK {
    index: u256,
  } 
}
```
so we can work on Token\<BAYC\> or Token\<PUNK\> with compile time type check instead of Token with potential missing runtime check.

happy to see how developers feel about different approaches, I think it's a tradeoff between ergonomics vs safety, unblocking for now

**Review by zekun000 — APPROVED** [review]

**Comment by davidiw** [comment]
> ncern of this is about type safety as we discussed.
> 
> the alternative way of doing this is by generics, and treating collection as Type, token as instance of one Type.

Agreed. I think for creative content and solutions that aren't likely to be entangled into DeFi, this substantially lowers the bar while still ensuring folks have a safe way to work with their NFTs and other related tokens.

**Comment by hariria** [comment]
@davidiw do we need to bump the Aptos npm package to `0.0.19` with these changes included?

**Pull request merged**